dipy/nn/synb0.py

```diff
@@ -0,0 +1,417 @@
+                             (77, 91, 77) for both inputs')
+
+        shape = b0.shape
+        dim = len(shape)
+
+        if dim == 3:
+            T1 = np.pad(T1, ((2, 1), (3, 2), (2, 1)), 'constant')
+            b0 = np.pad(b0, ((2, 1), (3, 2), (2, 1)), 'constant')
+            p99 = np.percentile(b0, 99)
+            T1 = self.__normalize(T1, 150, 0)
+            b0 = self.__normalize(b0, p99, 0)
+        else:
+            T1 = np.pad(T1, ((0, 0), (2, 1), (3, 2), (2, 1)), 'constant')
+            b0 = np.pad(b0, ((0, 0), (2, 1), (3, 2), (2, 1)), 'constant')
+            p99 = np.percentile(b0, 99, axis=(1, 2, 3))
+            for i in range(shape[0]):
+                T1[i] = self.__normalize(T1[i], 150, 0)
+                b0[i] = self.__normalize(b0[i], p99[i], 0)
+
+        if average:
+            mean_pred = np.zeros(shape+(5,), dtype=np.float32)
+            for i in range(5):
+                self.fetch_default_weights(i)
+                if dim == 3:
+                    if batch_size is not None:
+                        logger.warning('Batch size specified, but not used',
+                                       'due to the input not having \
+                                        a batch dimension')
+                    temp = np.expand_dims(np.stack([b0, T1], -1), 0)
+                    input_data = np.moveaxis(temp, 3, 1).astype(np.float32)
+                    prediction = self.__predict(input_data)
+                    prediction = self.__unnormalize(prediction, p99, 0)
+                    prediction = prediction[0, 2:-1, 2:-1, 3:-2, 0]
+                    prediction = np.moveaxis(prediction, 0, -1)
+
+                else:
+                    if batch_size is None:
+                        batch_size = 1
+                    temp = np.stack([b0, T1], -1)
+                    input_data = np.moveaxis(temp, 3, 1).astype(np.float32)
+                    prediction = np.zeros((shape[0], 80, 80, 96, 1),
+                                          dtype=np.float32)
+                    for batch_idx in range(batch_size, shape[0]+1, batch_size):
+                        temp_pred = self.__predict(input_data[:batch_idx])
+                        prediction[:batch_idx] = temp_pred
+                    remainder = np.mod(shape[0], batch_size)
+                    if remainder != 0:
+                        temp_pred = self.__predict(input_data[-remainder:])
+                        prediction[-remainder:] = temp_pred
+                    for j in range(shape[0]):
+                        temp = self.__unnormalize(prediction[j], p99[j], 0)
+                        prediction[j] = temp
+
+                    prediction = prediction[:, 2:-1, 2:-1, 3:-2, 0]
+                    prediction = np.moveaxis(prediction, 1, -1)
+
+                mean_pred[..., i] = prediction
+            prediction = np.mean(mean_pred, axis=-1)
+        else:
+            if dim == 3:
```

Comment:
Would it be simpler at the beginning to verify if 3D and to simply add an extra dimension to fake a batch size of 1?
This would simplify the whole thing from the start, no?

Reply:
It could. I'll look into it and see if that's possible.